src/Compilers/Core/Portable/InternalUtilities/OneOrMany.cs

```diff
@@ -37,19 +41,30 @@ public OneOrMany(ImmutableArray<T> many)
                 throw new ArgumentNullException(nameof(many));
             }
 
-            _one = default;
-            _many = many;
+            if (many is [var item])
+            {
+                _one = item;
+                _many = default;
```

Comment:
I understand this makes the internal invariants of the type a lot simpler but it does mean we essentially introduce allocations for this pattern where before there was no allocation: 

```csharp
var array = ImmutableArray.Create<int>(42);
var oneOrMany = new OneOrMany<int>(array);
// Before no alloc, after allocs 
oneOrMany.ToImmutable();
```

Are we concerned this could regress allocations? Don't have a guid intuition on how common this would be.

Reply:
Yes, potentially. I don't think it's common since we usually know at creation time that we only have a single element so we wouldn't create that array.